### PR TITLE
Update SIP-7 with the new form components

### PR DESCRIPTION
**Pull request by GuillaumeRx**
Updates the SIP 7 with the new form components

### CODE REVIEW

File: SIPS/sip-7.md [hunk 1]
```diff
@@ -84,13 +92,14 @@ interface HeadingComponent extends BaseComponent {
...
   type: 'text';
-  text: string;
+  value: string;
 }
 
 interface ButtonComponent extends BaseComponent {
   type: 'button';
   variant?: 'primary' | 'secondary';
-  text: string;
+  buttonType?: 'button' | 'submit';
```
Comment (Mrtenz) [inline SIPS/sip-7.md:101]:
```suggestion
  type?: 'button' | 'submit';
```

File: SIPS/sip-7.md [hunk 2]
```diff
@@ -65,12 +65,20 @@ This SIP describes a couple of components for the initial version of the UI:
 4. A button component, rendering the text as a HTML button (`button`), with an optional ID for handling user input.
     1. The button can have two variants: `primary` and `secondary`.
     2. If the variant is not specified, `primary` SHOULD be used.
+    3. The button can have two types: `button` and `submit`.
+    4. If the type is not specified, `button` SHOULD be used.
 5. A divider component, rendering a HTML thematic break (`hr`).
 6. A spacer component, pushing content around it to the top and bottom, similar to how `flex-grow` works.
 7. A spinner component, rendering a spinner to indicate loading data.
+8. A copyable component, rendering the value along a copy icon that copies the value in the clipboard.
+9. A Form component, which renders a list of `children` components and define a form, with an ID for handling user input.
```
Comment (Mrtenz) [inline SIPS/sip-7.md:74]:
```suggestion
9. A form component, which renders a list of `children` components to define a form, with an ID for handling user input.
```

File: SIPS/sip-7.md [hunk 3]
```diff
@@ -65,12 +65,20 @@ This SIP describes a couple of components for the initial version of the UI:
...
+10. An input component, rendering an HTML input (`input`) with an ID for handling user input.
+  1. The input can be of 4 different types: `text`, `password`, `number`, `search`
```
Comment (Mrtenz) [inline SIPS/sip-7.md:76]: What's the use case for `search` inputs?

File: SIPS/sip-7.md [hunk 4]
```diff
@@ -65,12 +65,20 @@ This SIP describes a couple of components for the initial version of the UI:
...
+  2. If the type is not specified, `text` SHOULD be used.
+  3. An aditional label MUST optionally be rendered if defined in the `label` field.
```
Comment (Mrtenz) [inline SIPS/sip-7.md:78]:
```suggestion
  3. An additional label MUST optionally be rendered if defined in the `label` field.
```

File: SIPS/sip-7.md [hunk 5]
```diff
@@ -106,14 +115,37 @@ interface SpinnerComponent extends BaseComponent {
...
+  type: 'form';
+  name: string;
+  children: Component[];
+}
+
+interface InputComponent extends BaseComponent {
+  type: 'input';
+  value?: string;
+  name: string;
+  inputType?: 'text' | 'password' | 'number' | 'search';
```
Comment (Mrtenz) [inline SIPS/sip-7.md:133]:
```suggestion
  type?: 'text' | 'password' | 'number' | 'search';
```

File: SIPS/sip-7.md [hunk 6]
```diff
@@ -198,6 +230,25 @@ interface SnapReadInterfaceRequest {
 type SnapReadInterfaceResponse = Json;
 ```
 
+#### `snap_getInterfaceState`
+
+The `snap_getInterfaceState` method lets Snaps retrive an existing user interface state. The params MUST be an object, with the following properties:
```
Comment (Mrtenz) [inline SIPS/sip-7.md:235]:
```suggestion
The `snap_getInterfaceState` method lets snaps retrieve an existing user interface state. The params MUST be an object, with the following properties:
```

File: SIPS/sip-7.md [hunk 7]
```diff
@@ -198,6 +230,25 @@ interface SnapReadInterfaceRequest {
...
+
+- `id` (`string`): The ID of the custom UI to read.
+
+This method returns the interface state.
```
Comment (Mrtenz) [inline SIPS/sip-7.md:239]: What is "interface state"?

**Review by Mrtenz — CHANGES_REQUESTED** [review]

File: SIPS/sip-7.md (hunk 1 above)
Comment (GuillaumeRx) [reply SIPS/sip-7.md:101]: we cannot use `type`. It's already used to determine what component is rendered

File: SIPS/sip-7.md (hunk 5 above)
Comment (GuillaumeRx) [reply SIPS/sip-7.md:133]: same as https://github.com/MetaMask/SIPs/pull/100#discussion_r1261163627

File: SIPS/sip-7.md (hunk 3 above)
Comment (GuillaumeRx) [reply SIPS/sip-7.md:76]: This is one of the available input types in the extension. I just allowed all the input types available on the `FormTextField` component that we use

File: SIPS/sip-7.md (hunk 7 above)
Comment (GuillaumeRx) [reply SIPS/sip-7.md:239]: The interfaces, AKA the UI has a state that stores the values of the statefull components like forms and inputs :)

Comment (Mrtenz) [reply SIPS/sip-7.md:239]: Can you clarify this in the SIP?

File: SIPS/sip-7.md [hunk 8]
```diff
@@ -84,13 +98,22 @@ interface HeadingComponent extends BaseComponent {
...
   type: 'button';
   variant?: 'primary' | 'secondary';
-  text: string;
+  buttonType: 'button'
+  value: string;
+  name?: string;
+}
+
+interface SubmitButtonComponent extends BaseComponent {
+  type: 'button';
```
Comment (Mrtenz) [inline SIPS/sip-7.md:113]:
```suggestion
  type: 'submitButton';
```

File: SIPS/sip-7.md [hunk 9]
```diff
@@ -84,13 +98,22 @@ interface HeadingComponent extends BaseComponent {
...
+  variant?: 'primary' | 'secondary';
```
Comment (Mrtenz) [inline SIPS/sip-7.md:114]: Do we need this for submit buttons? Shouldn't it just always be "primary"?

File: SIPS/sip-7.md [hunk 10]
```diff
@@ -106,14 +129,67 @@ interface SpinnerComponent extends BaseComponent {
...
+  inputType: 'text'
+  placeholder?: string;
+  label?: string;
+}
+
+interface PasswordInputComponent extends BaseComponent {
+  type: 'input';
+  value?: string;
+  name: string;
+  inputType: 'password'
```
Comment (Mrtenz) [inline SIPS/sip-7.md:156]:
```suggestion
```

File: SIPS/sip-7.md [hunk 11]
```diff
@@ -106,14 +129,67 @@ interface SpinnerComponent extends BaseComponent {
...
+}
+
+interface FormComponent extends BaseComponent {
+  type: 'form';
+  name: string;
+  children: Component[];
+}
+
+interface TextInputComponent extends BaseComponent {
+  type: 'input';
```
Comment (Mrtenz) [inline SIPS/sip-7.md:144]:
```suggestion
  type: 'textInput';
```

File: SIPS/sip-7.md [hunk 12]
```diff
@@ -106,14 +129,67 @@ interface SpinnerComponent extends BaseComponent {
...
+  value?: string;
+  name: string;
+  inputType: 'text'
+  placeholder?: string;
+  label?: string;
+}
+
+interface PasswordInputComponent extends BaseComponent {
+  type: 'input';
```
Comment (Mrtenz) [inline SIPS/sip-7.md:153]:
```suggestion
  type: 'passwordInput';
```

File: SIPS/sip-7.md [hunk 13]
```diff
@@ -106,14 +129,67 @@ interface SpinnerComponent extends BaseComponent {
...
+  value?: string;
+  name: string;
+  inputType: 'password'
+  placeholder?: string;
+  label?: string;
+}
+
+interface NumberInputComponent extends BaseComponent {
+  type: 'input';
```
Comment (Mrtenz) [inline SIPS/sip-7.md:162]:
```suggestion
  type: 'numberInput';
```

File: SIPS/sip-7.md [hunk 14]
```diff
@@ -106,14 +129,67 @@ interface SpinnerComponent extends BaseComponent {
...
+  value?: string;
+  name: string;
+  inputType: 'number'
```
Comment (Mrtenz) [inline SIPS/sip-7.md:165]:
```suggestion
```

File: SIPS/sip-7.md [hunk 15]
```diff
@@ -106,14 +129,67 @@ interface SpinnerComponent extends BaseComponent {
...
+  placeholder?: string;
+  label?: string;
+}
+interface SearchInputComponent extends BaseComponent {
+  type: 'input';
```
Comment (Mrtenz) [inline SIPS/sip-7.md:170]:
```suggestion
  type: 'searchInput';
```

File: SIPS/sip-7.md [hunk 16]
```diff
@@ -106,14 +129,67 @@ interface SpinnerComponent extends BaseComponent {
...
+  value?: string;
+  name: string;
+  inputType: 'search'
```
Comment (Mrtenz) [inline SIPS/sip-7.md:173]:
```suggestion
```

**Review by Mrtenz — CHANGES_REQUESTED** [review]

File: SIPS/sip-7.md (hunk 3 above)
Comment (Mrtenz) [reply SIPS/sip-7.md]: But it won't be handled any different compared to regular text inputs, right? I'd say we can just remove it.

File: SIPS/sip-7.md (hunk 9 above)
Comment (GuillaumeRx) [reply SIPS/sip-7.md:112]: People might want to have a secondary as submit and a primary for something else

Comment (Mrtenz) [reply SIPS/sip-7.md:112]: That seems like bad UX. 😅

Comment (GuillaumeRx) [reply SIPS/sip-7.md:112]: It's up to the snap developer IMO, it doesn't mean bad UX for all cases

File: SIPS/sip-7.md [hunk 17]
```diff
@@ -65,12 +65,24 @@ This SIP describes a couple of components for the initial version of the UI:
...
+4. A submit button component, rendering the text as a HTML button (`button`) of type `submit`, with an optional ID for handling user input.
+    1. The button can have two variants: `primary` and `secondary`.
+    2. If the variant is not specified, `primary` SHOULD be used.
 5. A divider component, rendering a HTML thematic break (`hr`).
 6. A spacer component, pushing content around it to the top and bottom, similar to how `flex-grow` works.
 7. A spinner component, rendering a spinner to indicate loading data.
+8. A copyable component, rendering the value along a copy icon that copies the value in the clipboard.
+9. A form component, which renders a list of `children` components to define a form, with an ID for handling user input.
+10. A text input component, rendering an HTML input (`input`) of type `text` with an ID for handling user input.
+  1. An aditional label MUST optionally be rendered if defined in the `label` field.
```
Comment (Montoya) [inline SIPS/sip-7.md:77]: MUST optionally?

**Review by Montoya — CHANGES_REQUESTED** [review]
Looks good, just wondering about "MUST optionally" in a few places.

**Review by Mrtenz — APPROVED** [review]

**Review by Montoya — APPROVED** [review]

**Pull request merged**